File: golem-common/src/schema/schema_type.rs
// Copyright 2024-2026 Golem Cloud
//
// Licensed under the Golem Source License v1.1 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://license.golem.cloud/LICENSE
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

use crate::schema::metadata::{MetadataEnvelope, TypeId};
use serde::{Deserialize, Serialize};

/// One schema type in the recursive in-memory representation.
///
/// Recursive positions hold owned, boxed children (rather than indices) so
/// consumers can walk and pattern-match the schema directly without chasing
/// indices into a flat node list. The WIT-shaped flat form is reconstructed
/// on demand by [`super::wit`].
///
/// Closed sum types come in two shapes that differ by how the decoder learns
/// which branch a value belongs to:
///
/// - [`SchemaType::Variant`] is a **carried-tag** sum: the value explicitly
///   carries its case (via [`super::VariantValuePayload::case`]). Zero-
///   inference decoding. Natural mapping for language-level algebraic data
///   types.
/// - [`SchemaType::Union`] is an **inferred-tag** sum: the value does not
///   carry its tag. Each branch declares a [`DiscriminatorRule`] and the
///   decoder picks the branch whose rule matches the raw value.
#[derive(Clone, Debug, PartialEq, Eq, Serialize, Deserialize)]
#[serde(tag = "kind", content = "value", rename_all = "kebab-case")]
pub enum SchemaType {
    /// Reference to a named definition in the enclosing
    /// [`super::SchemaGraph`].
    Ref(TypeId),

    // Primitives
    Bool,
    S8,
    S16,
    S32,
    S64,
    U8,
    U16,
    U32,
    U64,
    F32,
    F64,
    Char,
    String,

    // Structural composites
    Record {
        fields: Vec<NamedFieldType>,
    },
    Variant {
        cases: Vec<VariantCaseType>,
    },
    Enum {
        cases: Vec<String>,
    },
    Flags {
        flags: Vec<String>,
    },
    Tuple {
        elements: Vec<SchemaType>,
    },
    List {
        element: Box<SchemaType>,
    },
    FixedList {
        element: Box<SchemaType>,
        length: u32,
    },
    Map {
        key: Box<SchemaType>,
        value: Box<SchemaType>,
    },
    Option {
        inner: Box<SchemaType>,
    },
    Result(ResultSpec),

    // Rich semantic types
    Text(TextRestrictions),
    Binary(BinaryRestrictions),
    Path(PathSpec),
    Url(UrlRestrictions),
    Datetime,
    Duration,
    Quantity(QuantitySpec),

    // Discriminated union (closed, inferred-tag)
    Union(UnionSpec),

    // Capability nodes
    Secret(SecretSpec),
    QuotaToken(QuotaTokenSpec),

    // WASI P3 stubs (parseable only; no semantics yet).
    Future {
        inner: Option<Box<SchemaType>>,
    },
    Stream {
        inner: Option<Box<SchemaType>>,
    },
}

#[derive(Clone, Debug, PartialEq, Eq, Serialize, Deserialize)]
pub struct NamedFieldType {
    pub name: String,
    pub body: SchemaType,
    #[serde(default, skip_serializing_if = "MetadataEnvelope::is_empty")]
    pub metadata: MetadataEnvelope,
}

#[derive(Clone, Debug, PartialEq, Eq, Serialize, Deserialize)]
pub struct VariantCaseType {
    pub name: String,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub payload: Option<SchemaType>,
    #[serde(default, skip_serializing_if = "MetadataEnvelope::is_empty")]
    pub metadata: MetadataEnvelope,
}

#[derive(Clone, Debug, PartialEq, Eq, Serialize, Deserialize)]
pub struct ResultSpec {
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub ok: Option<Box<SchemaType>>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub err: Option<Box<SchemaType>>,
}

// --- Text / Binary ---

#[derive(Clone, Debug, Default, PartialEq, Eq, Serialize, Deserialize)]
pub struct TextRestrictions {
    /// Optional set of allowed BCP-47 language codes. `None` = unrestricted.
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub languages: Option<Vec<String>>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub min_length: Option<u32>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub max_length: Option<u32>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub regex: Option<String>,
}

#[derive(Clone, Debug, Default, PartialEq, Eq, Serialize, Deserialize)]
pub struct BinaryRestrictions {
    /// Optional set of allowed MIME types. `None` = unrestricted.
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub mime_types: Option<Vec<String>>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub min_bytes: Option<u32>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub max_bytes: Option<u32>,
}

// --- Path ---

#[derive(Copy, Clone, Debug, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "kebab-case")]
pub enum PathDirection {
    Input,
    Output,
    InOut,
}

#[derive(Copy, Clone, Debug, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "kebab-case")]
pub enum PathKind {
    File,
    Directory,
    Any,
}

#[derive(Clone, Debug, PartialEq, Eq, Serialize, Deserialize)]
pub struct PathSpec {
    pub direction: PathDirection,
    pub kind: PathKind,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub allowed_mime_types: Option<Vec<String>>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub allowed_extensions: Option<Vec<String>>,
}

// --- URL ---

#[derive(Clone, Debug, Default, PartialEq, Eq, Serialize, Deserialize)]
pub struct UrlRestrictions {
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub allowed_schemes: Option<Vec<String>>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub allowed_hosts: Option<Vec<String>>,
}

// --- Quantity ---

/// Fixed-point decimal value with unit: numeric value = `mantissa * 10^(-scale)`.
/// `unit` is a free-form string at the value level; the schema's
/// [`QuantitySpec`] constrains the accepted set.
#[derive(Clone, Debug, PartialEq, Eq, Serialize, Deserialize)]
pub struct QuantityValue {
    pub mantissa: i64,
    pub scale: i32,
    pub unit: String,
}

#[derive(Clone, Debug, PartialEq, Eq, Serialize, Deserialize)]
pub struct QuantitySpec {
    /// Canonical base unit (e.g., `"kg"`, `"m"`, `"s"`, `"B"`).
    pub base_unit: String,
    /// Suffixes accepted on input and rendered on output.
    #[serde(default, skip_serializing_if = "Vec::is_empty")]
    pub allowed_suffixes: Vec<String>,
    /// Optional inclusive range, expressed in canonical fixed-point form.
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub min: Option<QuantityValue>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub max: Option<QuantityValue>,
}

// --- Discriminated union ---

#[derive(Clone, Debug, PartialEq, Eq, Serialize, Deserialize)]
pub struct UnionSpec {
    pub branches: Vec<UnionBranch>,
}

#[derive(Clone, Debug, PartialEq, Eq, Serialize, Deserialize)]
pub struct UnionBranch {
    /// Logical branch name carried in
    /// [`super::UnionValuePayload::tag`] after the decoder resolves the
    /// branch; used by renderers, codegen, docs.
    pub tag: String,
    /// Branch body type. Must be compatible with the discriminator rule (see
    /// schema-construction validation rules in the design).
    pub body: SchemaType,
    /// Rule the decoder uses to pick this branch from a raw value.
    pub discriminator: DiscriminatorRule,
    #[serde(default, skip_serializing_if = "MetadataEnvelope::is_empty")]
    pub metadata: MetadataEnvelope,
}

/// How the decoder identifies that a value belongs to a given union branch.
#[derive(Clone, Debug, PartialEq, Eq, Serialize, Deserialize)]
#[serde(tag = "rule", content = "value", rename_all = "kebab-case")]
pub enum DiscriminatorRule {
    /// String value starts with this prefix.
    Prefix { prefix: String },
    /// String value ends with this suffix.
    Suffix { suffix: String },
    /// String value contains this substring.
    Contains { substring: String },
    /// String value matches this anchored regex.
    Regex { regex: String },
    /// Record-shaped value where the named field is present, and — if
    /// `literal` is set — has the given literal string value.
    FieldEquals(FieldDiscriminator),
    /// Record-shaped value where the named field is absent.
    FieldAbsent { field_name: String },
}

#[derive(Clone, Debug, PartialEq, Eq, Serialize, Deserialize)]
pub struct FieldDiscriminator {
    pub field_name: String,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub literal: Option<String>,
}

// --- Capability nodes ---

#[derive(Clone, Debug, Default, PartialEq, Eq, Serialize, Deserialize)]
pub struct SecretSpec {
    /// Optional categorisation (e.g., `"api-key"`, `"oauth-token"`).
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub category: Option<String>,
    #[serde(default, skip_serializing_if = "MetadataEnvelope::is_empty")]
    pub metadata: MetadataEnvelope,
}

#[derive(Clone, Debug, Default, PartialEq, Eq, Serialize, Deserialize)]
pub struct QuotaTokenSpec {
    /// Resource name this token covers (declared in the agent manifest).
    /// `None` = any resource permitted.
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub resource_name: Option<String>,
    #[serde(default, skip_serializing_if = "MetadataEnvelope::is_empty")]
    pub metadata: MetadataEnvelope,
}

File: sdks/scala/model/src/main/scala/golem/schema/SchemaType.scala
/*
 * Copyright 2024-2026 Golem Cloud
 *
 * Licensed under the Golem Source License v1.1 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://license.golem.cloud/LICENSE
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package golem.schema

/**
 * Recursive, in-memory mirror of a `golem:core/types@2.0.0` `schema-type-node`:
 * a structural [[SchemaTypeBody]] plus its [[MetadataEnvelope]].
 *
 * The WIT carrier (`schema-graph`) is flat-with-indices because WIT has no
 * recursive types. This recursive form is what the rest of the SDK works with;
 * [[golem.schema.wire]] converts to and from the flat carrier mechanically.
 */
final case class SchemaType(
  body: SchemaTypeBody,
  metadata: MetadataEnvelope = MetadataEnvelope.empty
)

/** A named field of a [[SchemaTypeBody.RecordType]]. */
final case class NamedFieldType(
  name: String,
  body: SchemaType,
  metadata: MetadataEnvelope = MetadataEnvelope.empty
)

/** A case of a [[SchemaTypeBody.VariantType]]. */
final case class VariantCaseType(
  name: String,
  payload: Option[SchemaType] = None,
  metadata: MetadataEnvelope = MetadataEnvelope.empty
)

/** A branch of a [[SchemaTypeBody.UnionType]] (closed, inferred-tag sum). */
final case class UnionBranch(
  tag: String,
  body: SchemaType,
  discriminator: DiscriminatorRule,
  metadata: MetadataEnvelope = MetadataEnvelope.empty
)

/** The structural body of a [[SchemaType]]. */
sealed trait SchemaTypeBody extends Product with Serializable

object SchemaTypeBody {
  // Reference to a named definition in the enclosing graph.
  final case class RefType(id: String) extends SchemaTypeBody

  // Primitives
  case object BoolType                                                       extends SchemaTypeBody
  final case class S8Type(restrictions: Option[NumericRestrictions] = None)  extends SchemaTypeBody
  final case class S16Type(restrictions: Option[NumericRestrictions] = None) extends SchemaTypeBody
  final case class S32Type(restrictions: Option[NumericRestrictions] = None) extends SchemaTypeBody
  final case class S64Type(restrictions: Option[NumericRestrictions] = None) extends SchemaTypeBody
  final case class U8Type(restrictions: Option[NumericRestrictions] = None)  extends SchemaTypeBody
  final case class U16Type(restrictions: Option[NumericRestrictions] = None) extends SchemaTypeBody
  final case class U32Type(restrictions: Option[NumericRestrictions] = None) extends SchemaTypeBody
  final case class U64Type(restrictions: Option[NumericRestrictions] = None) extends SchemaTypeBody
  final case class F32Type(restrictions: Option[NumericRestrictions] = None) extends SchemaTypeBody
  final case class F64Type(restrictions: Option[NumericRestrictions] = None) extends SchemaTypeBody
  case object CharType                                                       extends SchemaTypeBody
  case object StringType                                                     extends SchemaTypeBody

  // Structural composites
  final case class RecordType(fields: List[NamedFieldType])                    extends SchemaTypeBody
  final case class VariantType(cases: List[VariantCaseType])                   extends SchemaTypeBody
  final case class EnumType(cases: List[String])                               extends SchemaTypeBody
  final case class FlagsType(names: List[String])                              extends SchemaTypeBody
  final case class TupleType(elements: List[SchemaType])                       extends SchemaTypeBody
  final case class ListType(element: SchemaType)                               extends SchemaTypeBody
  final case class FixedListType(element: SchemaType, length: Int)             extends SchemaTypeBody
  final case class MapType(key: SchemaType, value: SchemaType)                 extends SchemaTypeBody
  final case class OptionType(element: SchemaType)                             extends SchemaTypeBody
  final case class ResultType(ok: Option[SchemaType], err: Option[SchemaType]) extends SchemaTypeBody

  // Rich semantic types
  final case class TextType(restrictions: TextRestrictions)     extends SchemaTypeBody
  final case class BinaryType(restrictions: BinaryRestrictions) extends SchemaTypeBody
  final case class PathType(spec: PathSpec)                     extends SchemaTypeBody
  final case class UrlType(restrictions: UrlRestrictions)       extends SchemaTypeBody
  case object DatetimeType                                      extends SchemaTypeBody
  case object DurationType                                      extends SchemaTypeBody
  final case class QuantityType(spec: QuantitySpec)             extends SchemaTypeBody

  // Discriminated union (closed, inferred-tag)
  final case class UnionType(branches: List[UnionBranch]) extends SchemaTypeBody

  // Capability nodes
  final case class SecretType(spec: SecretSpec)         extends SchemaTypeBody
  final case class QuotaTokenType(spec: QuotaTokenSpec) extends SchemaTypeBody

  // WASI P3 stubs (parseable only; no semantics yet)
  final case class FutureType(element: Option[SchemaType]) extends SchemaTypeBody
  final case class StreamType(element: Option[SchemaType]) extends SchemaTypeBody
}

/**
 * Compact constructors for [[SchemaType]]s (anonymous unless registered into a
 * graph via [[SchemaBuilder]]). Mirrors the TS SDK's `t` helpers.
 */
object t {
  import SchemaTypeBody._

  private def st(body: SchemaTypeBody): SchemaType = SchemaType(body)

  def ref(id: String): SchemaType                                = st(RefType(id))
  def bool: SchemaType                                           = st(BoolType)
  def s8: SchemaType                                             = s8(None)
  def s8(restrictions: Option[NumericRestrictions]): SchemaType  = st(S8Type(restrictions.flatMap(_.normalize)))
  def s16: SchemaType                                            = s16(None)
  def s16(restrictions: Option[NumericRestrictions]): SchemaType = st(S16Type(restrictions.flatMap(_.normalize)))
  def s32: SchemaType                                            = s32(None)
  def s32(restrictions: Option[NumericRestrictions]): SchemaType = st(S32Type(restrictions.flatMap(_.normalize)))
  def s64: SchemaType                                            = s64(None)
  def s64(restrictions: Option[NumericRestrictions]): SchemaType = st(S64Type(restrictions.flatMap(_.normalize)))
  def u8: SchemaType                                             = u8(None)
  def u8(restrictions: Option[NumericRestrictions]): SchemaType  = st(U8Type(restrictions.flatMap(_.normalize)))
  def u16: SchemaType                                            = u16(None)
  def u16(restrictions: Option[NumericRestrictions]): SchemaType = st(U16Type(restrictions.flatMap(_.normalize)))
  def u32: SchemaType                                            = u32(None)
  def u32(restrictions: Option[NumericRestrictions]): SchemaType = st(U32Type(restrictions.flatMap(_.normalize)))
  def u64: SchemaType                                            = u64(None)
  def u64(restrictions: Option[NumericRestrictions]): SchemaType = st(U64Type(restrictions.flatMap(_.normalize)))
  def f32: SchemaType                                            = f32(None)
  def f32(restrictions: Option[NumericRestrictions]): SchemaType = st(F32Type(restrictions.flatMap(_.normalize)))
  def f64: SchemaType                                            = f64(None)
  def f64(restrictions: Option[NumericRestrictions]): SchemaType = st(F64Type(restrictions.flatMap(_.normalize)))
  def char: SchemaType                                           = st(CharType)
  def string: SchemaType                                         = st(StringType)

  def record(fields: List[NamedFieldType]): SchemaType                       = st(RecordType(fields))
  def variant(cases: List[VariantCaseType]): SchemaType                      = st(VariantType(cases))
  def `enum`(cases: List[String]): SchemaType                                = st(EnumType(cases))
  def flags(names: List[String]): SchemaType                                 = st(FlagsType(names))
  def tuple(elements: List[SchemaType]): SchemaType                          = st(TupleType(elements))
  def list(element: SchemaType): SchemaType                                  = st(ListType(element))
  def fixedList(element: SchemaType, length: Int): SchemaType                = st(FixedListType(element, length))
  def map(key: SchemaType, value: SchemaType): SchemaType                    = st(MapType(key, value))
  def option(element: SchemaType): SchemaType                                = st(OptionType(element))
  def result(ok: Option[SchemaType], err: Option[SchemaType]): SchemaType    = st(ResultType(ok, err))
  def datetime: SchemaType                                                   = st(DatetimeType)
  def duration: SchemaType                                                   = st(DurationType)
  def secret(inner: SchemaType, category: Option[String] = None): SchemaType =
    st(SecretType(SecretSpec(inner, category)))

  def field(name: String, body: SchemaType): NamedFieldType                          = NamedFieldType(name, body)
  def variantCase(name: String, payload: Option[SchemaType] = None): VariantCaseType =
    VariantCaseType(name, payload)
}
